src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/integration/cas2/Cas2AssessmentTest.kt

```diff
@@ -0,0 +1,144 @@
+import org.junit.jupiter.api.Nested
+import org.junit.jupiter.api.Test
+import org.springframework.test.web.reactive.server.returnResult
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.api.model.Cas2Assessment
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.api.model.ServiceName
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.api.model.UpdateCas2Assessment
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.integration.IntegrationTestBase
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.integration.givens.`Given a CAS2 Assessor`
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.integration.givens.`Given a CAS2 User`
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.jpa.entity.Cas2ApplicationEntity
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.jpa.entity.Cas2AssessmentRepository
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.jpa.entity.NomisUserEntity
+import java.time.OffsetDateTime
+import java.util.UUID
+
+class Cas2AssessmentTest : IntegrationTestBase() {
+
+  @SpykBean
+  lateinit var realAssessmentRepository: Cas2AssessmentRepository
+
+  @AfterEach
+  fun afterEach() {
+    // SpringMockK does not correctly clear mocks for @SpyKBeans that are also a @Repository, causing mocked behaviour
+    // in one test to show up in another (see https://github.com/Ninja-Squad/springmockk/issues/85)
+    // Manually clearing after each test seems to fix this.
+    clearMocks(realAssessmentRepository)
+  }
+
+  @Nested
+  inner class MissingJwt {
+    @Test
+    fun `updating an assessment without JWT returns 401`() {
+      webTestClient.put()
+        .uri("/cas2/assessments/de6512fc-a225-4109-bdcd-86c6307a5237")
+        .exchange()
+        .expectStatus()
+        .isUnauthorized
+    }
+  }
+
+  @Nested
+  inner class ControlsOnExternalUsers {
+    @Test
+    fun `updating an assessment is forbidden to external users who are not Assessors`() {
+      val jwt = jwtAuthHelper.createClientCredentialsJwt(
+        username = "username",
+        authSource = "auth",
+        roles = listOf("ROLE_CAS2_ADMIN"),
+      )
+
+      webTestClient.post()
+        .uri("/cas2/assessments/de6512fc-a225-4109-bdcd-86c6307a5237")
+        .header("Authorization", "Bearer $jwt")
+        .exchange()
+        .expectStatus()
+        .isForbidden
+    }
+  }
+
+  @Nested
+  inner class ControlsOnInternalUsers {
+    @Test
+    fun `updating an application is forbidden to nomis users`() {
+      val jwt = jwtAuthHelper.createClientCredentialsJwt(
+        username = "username",
+        authSource = "nomis",
+        roles = listOf("ROLE_POM"),
+      )
+
+      webTestClient.put()
+        .uri("/cas2/assessments/de6512fc-a225-4109-bdcd-86c6307a5237")
+        .header("Authorization", "Bearer $jwt")
+        .exchange()
+        .expectStatus()
+        .isForbidden
+    }
+  }
+
+  @Nested
+  inner class PutToUpdate {
+    @Test
+    fun `assessors create note returns 201`() {
```

Comment:
Are there some unhappy path tests too?

validationResult NotFound - throw
validationResult Unauthorised - throw

updatedAssessment GeneralValidationError - throw
updatedAssessment FieldValidationError - throw
updatedAssessment ConflictError - throw

Reply:
we've tested Unauthorised in the above tests, which returns Forbidden.
Good spot that `NotFound` isn't tested, but I think I will put that in the Service unit tests - I'm not entirely sure the others warrant an expensive integration test, from my understanding the validation/conflict errors are not actually being triggered by the Service method here, we just have to account for them to fulfil the `when` block.